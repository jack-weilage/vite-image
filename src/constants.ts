import type { FormatEnum } from 'sharp'
import type { ColorspaceEnum, PluginConfig } from '../types'

import Schema from 'validate'

export const DEFAULT_PLUGIN_CONFIG: PluginConfig = {
    include: '**/*.{heic,heif,avif,jpeg,jpg,png,tiff,webp,gif}?*',
    exclude: '',
    deliminator: ',',
    transformers: [],

    default_exports: [ 'src', 'aspect', 'width', 'height', 'format' ]
}

export const INPUT_FORMATS: (keyof FormatEnum)[] = [
    'avif',
    'dz',
    'fits',
    'gif',
    'heif',
    'input',
    'jpeg',
    'jpg',
    'magick',
    'openslide',
    'pdf',
    'png',
    'ppm',
    'raw',
    'svg',
    'tif',
    'tiff',
    'v',
    'webp'
]

/** @see https://github.com/libvips/libvips/blob/41cff4e9d0838498487a00623462204eb10ee5b8/libvips/iofuncs/enumtypes.c#L776-L796 */
export const INPUT_COLORSPACES: (keyof typeof ColorspaceEnum)[] = [
    // 'error',
    // 'multiband',
    'b-w',
    'histogram',
    'xyz',
    'lab',
    'cmyk',
    'labq',
    'rgb',
    'cmc',
    'lch',
    'labs',
    'srgb',
    'yxy',
    'fourier',
    'rgb16',
    'grey16',
    'matrix',
    'scrgb',
    'hsv'
    // 'last'
]

/* c8 ignore next 4 */
export const DEV_PREFIX = '/@image/'
export const DEV_REGEX = new RegExp(`^${DEV_PREFIX}([a-z0-9]{40})$`)
export const BUILD_PREFIX = '__VITE_IMAGE_ASSET__'
export const BUILD_REGEX = new RegExp(`${BUILD_PREFIX}([a-z0-9]{8})`, 'g')

export const CONFIG_SCHEMA = new Schema({
    include: { type: String },
    exclude: { type: String },
    deliminator: { type: String },
    transformers: [{
        name: { type: String, required: true },
        matcher: { type: Function, required: true },
        transform: { type: Function, required: true }
    }],
    default_exports: [{ type: String }]
}, { strict: true })