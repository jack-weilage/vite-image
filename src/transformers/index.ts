import format from './format'

import resize from './resize'

import rotate from './rotate'
import flip from './flip'
import flop from './flop'
import sharpen from './sharpen'
import median from './median'
import blur from './blur'
import gamma from './gamma'
import negate from './negate'
import normalize from './normalize'
import threshold from './threshold'
import modulate from './modulate'

import tint from './tint'
import grayscale from './grayscale'
import colorspace from './colorspace'

import metadata from './metadata'


export default [
    format,
    resize,
    rotate, flip, flop, sharpen, median, blur, gamma, negate, normalize, threshold, modulate,
    tint, grayscale, colorspace,
    metadata
]