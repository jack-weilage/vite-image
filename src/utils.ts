import type { ImageConfig, Transformer, PluginConfig } from '../types'
import type { Sharp, Metadata } from 'sharp'

import { type BinaryLike, createHash } from 'crypto'
import { basename, extname } from 'path'
import { CONFIG_SCHEMA } from './constants'

/** Limit a number to between min and max. */
export const minmax = (num: number, min: number, max: number) => Math.max(Math.min(num, max), min)
/** 
 * Create a SHA1 hash from a string.
 * 
 * SHA256 is more secure, but speed matters more here.
*/
export const create_hash = (str: BinaryLike): string => createHash('sha1').update(str).digest('hex')
/** Extract the name of a file, without its extension. */
export const filename = (path: string) => basename(path, extname(path))
/** De-duplicate an array. */
export const dedupe = <T>(arr: T[]) => [ ...new Set(arr) ]

/** Create a `Partial` of any object. */
export function copy_only_keys<T>(obj: T, keys: (keyof T)[]): Partial<T> {
    const partial = {} as Partial<T>
    for (const key of keys)
        partial[key] = obj[key]

    return partial
}

/** Validate and combine a user config and default config. */
export function parse_config(user_config: Partial<PluginConfig>, default_config: PluginConfig): PluginConfig
{
    const config = Object.assign({ ...default_config }, user_config)

    const errors = CONFIG_SCHEMA.validate(config)
    if (errors.length !== 0)
        throw new AggregateError(errors)
    
    return config
}

/** Coerces values to string | number | boolean. */
const format_value = (val: string) => {
    if (val === '' || val === 'true')
        return true

    if (val === 'false')
        return false
    
    if (!Number.isNaN(+val))
        return +val

    return val
}

//TODO: Handle `!blur` as `blur=false`
/** Create unique configs from arrays of values. */
export function create_configs(params: URLSearchParams, deliminator: string): ImageConfig[]
{
    const aggregated = {} as Record<string, (string | number | boolean)[]>
    for (const key of dedupe([ ...params.keys() ]))
    {
        // Get every occurrence of the key, then split by the deliminator.
        const values = params.getAll(key)
            .flatMap(value => value.split(deliminator))
            .map(format_value)

        aggregated[key] = dedupe(values)
    }
    
    const keys = Object.keys(aggregated)
    const values = Object.values(aggregated)

    // If there is only one key, the code below won't work.
    if (keys.length === 1)
        return values[0].map(value => ({ [keys[0]]: value }))
    
    // Create an array of unique configs.
    // Each value will have the index of the corresponding keys.
    const groups = values
        //@ts-expect-error: This reducer will take in (string | number | boolean)[] and return (string | number | boolean)[][]
        .reduce((acc, cur) => acc.flatMap(a => cur.map(b => [ a, b ].flat()))) as any as (string | number | boolean)[][]
    
    const configs = [] as ImageConfig[]
    for (const group of groups) {
        const config = {} as Record<string, string | number | boolean>

        // For every value, assign the corresponding key.
        for (let i = 0; i < keys.length; i++)
            config[keys[i]] = group[i]
        
        configs.push(config)
    }

    return configs
}

/** Apply all transformers to an image. */
export function apply_transformers(image: Sharp, metadata: Metadata, config: ImageConfig, transformers: Transformer[])
{
    const applied_transformers = []

    for (const { matcher, transform, name } of transformers)
    {
        if (!matcher(config))
            continue
    
        applied_transformers.push(name)
        image = transform(image, config, metadata)
    }

    return { image, applied_transformers }
}