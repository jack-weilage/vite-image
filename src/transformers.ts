import type { Sharp, FormatEnum } from 'sharp'
import type { Transformer } from '../types'


import { INPUT_FORMATS } from './constants'
import { includes } from './utils'


const resize = (img: Sharp, width: number, height: number) => 
    img.resize(Math.round(Math.max(width, 1)), Math.round(Math.max(height, 1)), { fit: 'fill' })

export default [
    // Output
    {
        name: 'format',
        matcher: (config) => includes(INPUT_FORMATS, config['format']),
        transform: (img, config) => img.toFormat(config['format'] as keyof FormatEnum)
    },
    // Resize
    {
        name: 'resize',
        matcher: (config) => {
            const { width, height } = config

            const width_valid = typeof width === 'number' && !isNaN(width)
            const height_valid = typeof height === 'number' && !isNaN(height)

            return width_valid && height_valid ||       // Both width and height are supplied
                width_valid && height === undefined ||  // Only width is supplied
                width === undefined && height_valid     // Only height is supplied
        },
        transform: (img, config, metadata) => {
            const { width, height } = config

            // Check this first as we don't need metadata.
            if (width !== undefined && height !== undefined)
                return resize(img, width, height)

            // Everything after this will need metadata width/height
            if (!metadata.width || !metadata.height)
                return
            

            if (width && height === undefined)
                return resize(img, width, width / metadata.width * metadata.height)

            if (width === undefined && height)
                return resize(img, height / metadata.height * metadata.width, height)
                
            throw new Error('This shouldn\'t happen. How did you get here?')
        }
    },
    // Image Operations
    {
        name: 'rotate',
        matcher: (config) => typeof config['rotate'] === 'number',
        transform: (img, config) => img.rotate(config['rotate'])
    },
    {
        name: 'flip',
        matcher: (config) => typeof config['flip'] === 'boolean',
        transform: (img, config) => img.flip(config['flip'])
    },
    {
        name: 'flop',
        matcher: (config) => typeof config['flop'] === 'boolean',
        transform: (img, config) => img.flop(config['flop'])
    },
    {
        name: 'sharpen',
        matcher: (config) => config['sharpen'] === true || typeof config['sharpen'] === 'number',
        transform: (img, config) => img.sharpen(config['sharpen'] === true ? undefined : { sigma: config['sharpen'] as number })
    },
    {
        name: 'median',
        matcher: (config) => config['median'] === true || typeof config['sharpen'] === 'number',
        transform: (img, config) => img.median(config['median'] === true ? undefined : config['median'])
    },
    {
        name: 'blur',
        matcher: (config) => config['blur'] === true || typeof config['blur'] === 'number',
        transform: (img, config) => img.blur(config['blur'] === true ? undefined : Math.max(config['blur'] as number, 0.3))
    },
    {
        name: 'gamma',
        matcher: (config) => config['gamma'] === true || typeof config['gamma'] === 'number',
        transform: (img, config) => img.gamma(config['gamma'] === true ? undefined : Math.min(Math.max(config['gamma'] as number, 1), 3))
    },
    {
        name: 'negate',
        matcher: (config) => typeof config['negate'] === 'boolean',
        transform: (img, config) => img.negate(config['negate'])
    },
    {
        name: 'normalize',
        matcher: (config) => typeof config['normalize'] === 'boolean',
        transform: (img, config) => img.normalize(config['normalize'])
    },
    {
        name: 'threshold',
        matcher: (config) => config['threshold'] === true || typeof config['threshold'] === 'number',
        transform: (img, config) => img.threshold(config['threshold'] === true ? undefined : Math.min(Math.max(config['threshold'] as number, 0), 255))
    },
    {
        name: 'modulate',
        // Ugly ugly ugly. All options are either undefined or number, but at least one option must be number.
        matcher: (config) => (typeof config['brightness'] === 'undefined' || typeof config['brightness'] === 'number')
        && (typeof config['saturation'] === 'undefined' || typeof config['saturation'] === 'number')
        && (typeof config['hue'] === 'undefined' || typeof config['hue'] === 'number')
        && (typeof config['lightness'] === 'undefined' || typeof config['lightness'] === 'number')
        && ((typeof config['brightness'] ?? config['saturation'] ?? config['hue'] ?? config['lightness']) === 'number'),
        transform: (img, config) => img.modulate({
            brightness: config['brightness'],
            saturation: config['saturation'],
            hue: config['hue'],
            lightness: config['lightness']
        })
    }
] as Transformer[]