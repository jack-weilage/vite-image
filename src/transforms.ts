import type { FormatEnum } from 'sharp'
import type { Transformer } from '../types'


import { INPUT_FORMATS } from './constants'
import { includes } from './utils'

export default [
    {
        name: 'resize',
        matcher: (config) => {
            const width = config['width']
            const height = config['height']

            const width_valid = typeof width === 'number' && !isNaN(width)
            const height_valid = typeof height === 'number' && !isNaN(height)

            return width_valid && height_valid ||       // Both width and height are supplied
                width_valid && height === undefined ||  // Only width is supplied
                width === undefined && height_valid     // Only height is supplied
        },
        transform: (img, config, metadata) => {
            const width = config['width']
            const height = config['height']

            // Check this first as we don't need metadata.
            if (width !== undefined && height !== undefined)
                return img.resize(width, height)

            // Everything after this will need metadata width/height
            if (!metadata.width || !metadata.height)
                return
            
            
            if (width && height === undefined)
                return img.resize(width, width / metadata.width * metadata.height)

            if (width === undefined && height)
                return img.resize(height / metadata.height * metadata.width, height)
                
            throw new Error('This shouldn\'t happen. How did you get here?')
        }
    },
    {
        name: 'format',
        matcher: (config) => includes(INPUT_FORMATS, config['format']),
        transform: (img, config) => img.toFormat(config['format'] as keyof FormatEnum)
    },
    {
        name: 'blur',
        matcher: (config) => config['blur'] === true || typeof config['blur'] === 'number',
        transform: (img, config) => img.blur(config['blur'] === true ? 0.3 : Math.max(config['blur'] as number, 0.3))
    }
] as Transformer[]