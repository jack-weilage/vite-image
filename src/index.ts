import type { Plugin, ResolvedConfig } from 'vite'
import type { Digest, DigestEntry, PluginConfig, InternalImage, OutputImage, Transformer } from '../types'

import { BUILD_PREFIX, DEFAULT_CONFIG, DEV_PREFIX } from './constants'
import transforms from './transforms'
import { apply_transforms, copy_only_keys, create_configs, create_hash, dedupe, filename, params_to_obj, parse_config } from './utils'

import { createFilter, dataToEsm } from '@rollup/pluginutils'
import MagicString from 'magic-string'
import sharp from 'sharp'
import { join } from 'path'


export type { PluginConfig, Transformer }

export default function image(user_config: Partial<PluginConfig> = {}): Plugin {
    const plugin_config: PluginConfig = parse_config(user_config, DEFAULT_CONFIG)

    const digested = new Map() as Digest
    const filter = createFilter(plugin_config.include, plugin_config.exclude)

    let viteConfig: ResolvedConfig

    return {
        name: 'image',
        enforce: 'pre',

        configResolved(config) { viteConfig = config },

        async load(id: string) {
            if (!filter(id))
                return null

            // `pathToFileURL` should be used here, but it doesn't parse like a normal url. Should be fine?
            const url = new URL(join(viteConfig.publicDir, id), 'file://')
            const base_img = sharp(url.pathname)
                .withMetadata()
            
            // Deal with output meta tags here so it can be removed from url.
            const exports = dedupe([
                ...plugin_config.default_exports,
                ...(url.searchParams.get('export') ?? '').split(plugin_config.deliminator) as (keyof OutputImage)[]
            ]).filter(Boolean)

            // If nothing is going to be output, why even process the image? This currently won't happen, as the defaults can't be overwritten.
            if (exports.length === 0)
            {
                if (this.meta.watchMode)
                    console.warn('Image', url.toString(), 'did not export any metadata.')
                
                return null
            }

            // Remove `meta` from search params to prevent having to deal with it later.
            url.searchParams.delete('meta')

            const images = [] as InternalImage[]
            for (const config of create_configs(params_to_obj(url.searchParams, plugin_config.deliminator))) {
                const hash = create_hash(url.toString() + JSON.stringify(config))

                // If we've already processed this exact image/config...
                if (digested.has(hash)) {
                    // Just grab what we already have from the digest.
                    images.push((digested.get(hash) as DigestEntry).data)
                    continue
                }

                const { img, is_transformed } = await apply_transforms(base_img.clone(), config, [...plugin_config.transformers, ...transforms])

                // If the image didn't match a transformer, it shouldn't be processed
                if (!is_transformed)
                    continue
                // Convert the transformed image to a buffer. We only need the buffer for build mode, but always need the metadata.
                //? Is there a way to get a transformed image's metadata without waiting for buffer?
                const { info, data: source } = await img.toBuffer({ resolveWithObject: true })

                // If we're in dev mode, we should supply an actual url here.
                let src = DEV_PREFIX + hash
                // If we're not in dev mode...
                if (!this.meta.watchMode) {
                    const name = `${filename(url.pathname)}.${info.format}`

                    const handle = this.emitFile({ name, source, type: 'asset' })
                    // We should make this path recognizable for `renderChunk`.
                    src = BUILD_PREFIX + handle
                }

                const data = {
                    ...info,
                    aspect: info.width / info.height,
                    src,
                }

                digested.set(hash, { img, data })
                images.push(data)
            }
            // If every config was skipped, we shouldn't change the output
            if (images.length === 0)
                return null

            const data: OutputImage[] = images.map(img => copy_only_keys(img, exports))
            return dataToEsm(data)
        },
        configureServer(server) {
            server.middlewares.use((req, res, next) => {
                if (!req.url)
                    return next()

                const result = new RegExp(`^${DEV_PREFIX}(.*)$`).exec(req.url)

                // If the path does not match our regex, or we haven't digested this image, skip.
                if (!result || !digested.has(result[1]))
                    return next()


                return (digested.get(result[1]) as DigestEntry).img
                    .clone()
                    .pipe(res)
            })
        },
        /**
         * Code shamelessly stolen from `vite-imagetools`.
         * renderChunk runs on every chunk of code vite is rendering.
         * This function replaces every occurrence of a flagged url with a proper URL.
         */
        renderChunk(code) {
            const regex = new RegExp(`${BUILD_PREFIX}([a-z\\d]{8})`, 'g')

            let match: RegExpExecArray | null
            let s: MagicString | undefined

            while (match = regex.exec(code)) {
                s = s || new MagicString(code)
                const [full, hash] = match

                // Starting at the index of the match, replace the length of the match with 
                s.overwrite(match.index, match.index + full.length, viteConfig.base + this.getFileName(hash))
            }

            if (s) {
                return {
                    code: s.toString(),
                    map: viteConfig.build.sourcemap ? s.generateMap({ hires: true }) : null
                }
            }

            return null
        }
    }
}