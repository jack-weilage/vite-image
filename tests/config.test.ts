import type { PluginConfig } from '../types'

import { parse_plugin_config } from '../src/utils'
import { it, expect } from 'vitest'

it.each([
    { test: true },
    { include: 10 },
    { transformers: [ 'foo', 'bar' ] }
])('does not apply invalid config: %s', (input) => {
    // This seems to be the best way to match an AggregateError.
    try {
        //@ts-expect-error: The config _should_ be invalid here.
        parse_plugin_config(input)
    } catch (error) {
        expect(error).toMatchSnapshot()
    }
})

it.each([
    undefined,
    {  },
    { include: '**/*' },
    { include: '**/*', exclude: '**/*.jpeg' },
    { 
        transformers: [{
            name: 'noop',
            matcher: () => true,
            transform: (img: any) => img
        }]
    }
])('affects custom imports: %s', (input) => {
    expect(parse_plugin_config(input as Partial<PluginConfig>)).toMatchSnapshot()
})