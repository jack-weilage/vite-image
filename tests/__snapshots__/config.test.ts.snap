// Vitest Snapshot v1

exports[`affects custom imports: { include: '**/*' } 1`] = `
{
  "default_exports": [
    "src",
    "aspect",
    "width",
    "height",
    "format",
  ],
  "deliminator": ",",
  "exclude": "",
  "include": "**/*",
  "post_process": [Function],
  "transformers": [],
}
`;

exports[`affects custom imports: { include: '**/*', exclude: '**/*.jpeg' } 1`] = `
{
  "default_exports": [
    "src",
    "aspect",
    "width",
    "height",
    "format",
  ],
  "deliminator": ",",
  "exclude": "**/*.jpeg",
  "include": "**/*",
  "post_process": [Function],
  "transformers": [],
}
`;

exports[`affects custom imports: { transformers: [Array] } 1`] = `
{
  "default_exports": [
    "src",
    "aspect",
    "width",
    "height",
    "format",
  ],
  "deliminator": ",",
  "exclude": "",
  "include": "**/*.{heic,heif,avif,jpeg,jpg,png,tiff,webp,gif}?*",
  "post_process": [Function],
  "transformers": [
    {
      "matcher": [Function],
      "name": "noop",
      "transform": [Function],
    },
  ],
}
`;

exports[`affects custom imports: {} 1`] = `
{
  "default_exports": [
    "src",
    "aspect",
    "width",
    "height",
    "format",
  ],
  "deliminator": ",",
  "exclude": "",
  "include": "**/*.{heic,heif,avif,jpeg,jpg,png,tiff,webp,gif}?*",
  "post_process": [Function],
  "transformers": [],
}
`;

exports[`affects custom imports: undefined 1`] = `
{
  "default_exports": [
    "src",
    "aspect",
    "width",
    "height",
    "format",
  ],
  "deliminator": ",",
  "exclude": "",
  "include": "**/*.{heic,heif,avif,jpeg,jpg,png,tiff,webp,gif}?*",
  "post_process": [Function],
  "transformers": [],
}
`;

exports[`does not apply invalid config: { include: 10 } 1`] = `[Error: Error: include must be of type String.]`;

exports[`does not apply invalid config: { test: true } 1`] = `[Error: Error: test is not allowed.]`;

exports[`does not apply invalid config: { transformers: [Array] } 1`] = `[Error: Error: transformers.0 must be of type Object.,Error: transformers.1 must be of type Object.,Error: transformers.0.name is required.,Error: transformers.1.name is required.,Error: transformers.0.matcher is required.,Error: transformers.1.matcher is required.,Error: transformers.0.transform is required.,Error: transformers.1.transform is required.]`;
